### PR TITLE
MMDGroup class

### LINKED ISSUE #16: Implement a Group object which allows searching for group-members
We need a method for searching among members of a specific group (i.e. https://vocab.met.no/mmd/Instrument) to replace the pythesint lookup

### PR DESCRIPTION
**Pull request by johtoblan**
**Summary**: Implements MMDGroup class to facilitate searching for elements in a given group hierarchy

**Related issue**:

Closes #16 

**Suggested reviewer(s)**: @vkbo 

**Reviewer checklist**:

* [x] The headers of all files contain a reference to the repository license
* [x] 100% test coverage of new code - meaning:
  * [x] The overall test coverage increased or remained the same as before
  * [x] Every function is accompanied with a test suite
  * [x] Tests are both positive (testing that the function work as intended with valid data) and negative (testing that the function behaves as expected with invalid data, e.g., that correct exceptions are thrown)
  * [x] Functions with optional arguments have separate tests for all options
* [x] Examples are supported by doctests
* [x] All tests are passing
* [x] All names (e.g., files, classes, functions, variables) are explicit
* [ ] Documentation (as docstrings) is complete and understandable

The checklist is based on the S-ENDA conventions and definition of done (see [General Conventions](https://s-enda-documentation.readthedocs.io/en/latest/general_conventions.html)). The above points are not necessarily relevant to all contributions. In that case, please add a short explanation to help the reviewer.

### CODE REVIEW

**Comment by vkbo** [comment]
Please make an issue task for this as well. Request from @mortenwh.

File: metvocab/mmdgroup.py [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+                    uri = member.get("uri")
+                    cache = DataCache()
+                    data = cache.get_vocab(self._voc_id, uri)
+                    data = self._get_concept_dictionary(data, uri)
+
+                    self._concepts.update({member.get("uri"): data})
+
+        return
+
+    def search(self, name):
```
Comment (vkbo) [inline metvocab/mmdgroup.py:48]: Docstring missing

File: metvocab/mmdgroup.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        self._voc_id = voc_id
+        self._uri = uri
+
+        self._cache = DataCache()
+        self._concepts = {}
+
+        self._data = self._cache.get_vocab(voc_id, uri)
+
+        # Populate _concepts with data-dictionary for members of uri-group
+        for graph in self._data.get("graph", []):
```
Comment (vkbo) [inline metvocab/mmdgroup.py:36]: Didn't we agree to populate data in a separate function?

File: metvocab/mmdgroup.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+            return value.get("value", None)
+        return None
+
+    def _get_resource(self, concept, label):
+        """Helper function for search method"""
+        value = concept.get(label)
+        if isinstance(value, dict):
+            return value.get("uri", "")
+        elif isinstance(value, list):
+            return value[0].get("uri", "")
```
Comment (vkbo) [inline metvocab/mmdgroup.py:91]: Don't assume the list has length > 0.

File: tests/test_mmdgroup.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
+"""
+MetVocab : MMD Vocabulary Class Tests
+=====================================
```
Comment (vkbo) [inline tests/test_mmdgroup.py:3]: Update module docstring title

File: tests/test_mmdgroup.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+import os
+import json
+import pytest
+
+from metvocab.cache import DataCache
+from metvocab.mmdgroup import MMDGroup
+
+
+@pytest.mark.live
+def testLiveMMDGroup_Init():
```
Comment (vkbo) [inline tests/test_mmdgroup.py:29]: docstring

File: tests/test_mmdgroup.py [hunk 6]
```diff
@@ -0,0 +1,177 @@
...
+    group = MMDGroup("mmd", "https://vocab.met.no/mmd/Platform")
+    assert len(group._concepts) > 0
+    group = MMDGroup("mmd", "https://vocab.met.no/mmd/NotAGroup")
+    assert len(group._concepts) == 0
+
+# END Test testLiveMMDGroup_Init
+
+
+@pytest.mark.live
+def testLiveMMDGroup_Search():
```
Comment (vkbo) [inline tests/test_mmdgroup.py:39]: docstring

File: tests/test_mmdgroup.py [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+    assert group.search("MockSat") != modis_dict
+    assert group.search("MockSat") != olci_dict
+
+# END Test testLiveMMDGroup_Init
+
+
+@pytest.mark.core
+def testCoreMMDGroup_Init(filesDir, monkeypatch):
+    """Tests initialisation of vocabulary against local files"""
+
```
Comment (vkbo) [inline tests/test_mmdgroup.py:69]: remove additional line break

File: tests/test_mmdgroup.py [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+        assert len(group._concepts) > 0
+        group = MMDGroup("mmd", "https://vocab.met.no/mmd/NotAGroup")
+        assert len(group._concepts) == 0
+
+# END TEST testCoreMMDGroup_Init
+
+
+@pytest.mark.core
+def testCoreMMDGroup_Search(filesDir):
+    """Tests search of vocabulary against local files"""
```
Comment (vkbo) [inline tests/test_mmdgroup.py:92]: Remove redundant empty line

File: tests/test_mmdgroup.py [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+
+    json_group_path = os.path.join(filesDir, "Instrument.json")
+    json_modis_path = os.path.join(filesDir, "Instrument", "MODIS.json")
+    json_olci_path = os.path.join(filesDir, "Instrument", "OLCI.json")
+
+    with open(json_group_path, mode="r", encoding="utf-8") as infile:
+        group_data = json.load(infile)
```
Comment (vkbo) [inline tests/test_mmdgroup.py:99]: Hmm, maybe just add a `readJson` function int the `tests/tools.py` file?

File: tests/test_mmdgroup.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+    assert group.search("MODIS") != olci_dict
+
+    assert group.search("Ocean and Land Colour Imager") == olci_dict
+    assert group.search("Ocean and Land Colour Imager") != modis_dict
+
+    assert group.search("MockSat") == {}
+    assert group.search("MockSat") != modis_dict
+    assert group.search("MockSat") != olci_dict
+
+
```
Comment (vkbo) [inline tests/test_mmdgroup.py:137]: Remove second empty line

File: tests/test_mmdgroup.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+# END TEST testCoreMMDGroup_Search
+
```
Comment (vkbo) [inline tests/test_mmdgroup.py:139]: Add one more empty line

File: tests/test_mmdgroup.py [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+def testCoreMMDGroup_GetLabel(monkeypatch):
```
Comment (vkbo) [inline tests/test_mmdgroup.py:140]: Add docstring
Add @pytest.mark.core

File: tests/test_mmdgroup.py [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+    concept = {"label": "value"}
+    concept_with_dict = {"label": {"value": "inner_value"}}
+
+    assert group._get_label(concept, "label") == "value"
+    assert group._get_label(concept, "value") is None
+
+    assert group._get_label(concept_with_dict, "label") == "inner_value"
+    assert group._get_label(concept_with_dict, "value") is None
+
+
```
Comment (vkbo) [inline tests/test_mmdgroup.py:154]: Remove additional empty line

File: tests/test_mmdgroup.py [hunk 14]
```diff
@@ -0,0 +1,177 @@
...
+# END TEST testCoreMMDGroup_GetLabel
+
```
Comment (vkbo) [inline tests/test_mmdgroup.py:156]: Add second blank line

File: tests/test_mmdgroup.py [hunk 15]
```diff
@@ -0,0 +1,177 @@
...
+def testCoreMMDGroup_GetResource(monkeypatch):
```
Comment (vkbo) [inline tests/test_mmdgroup.py:157]: Add docstring
Add @pytest.mark.core

File: tests/test_mmdgroup.py [hunk 16]
```diff
@@ -0,0 +1,177 @@
...
+    assert group._get_resource(concept_dict, "resource") != "SomethingElse"
+
+    assert group._get_resource(concept_with_list, "resource") == "https://vocab.met.no"
+    assert group._get_resource(concept_with_list, "resource") != "SomethingElse"
+
+    assert group._get_resource(concept_with_dict, "resource") == "https://vocab.met.no"
+    assert group._get_resource(concept_with_dict, "resource") != "SomethingElse"
+
+    assert group._get_resource(invalid_concept, "resource") == ""
+# END TEST testCoreMMDGroup_GetResource
```
Comment (vkbo) [inline tests/test_mmdgroup.py:177]: Add empty line above

**Review by vkbo — CHANGES_REQUESTED** [review]
Minor things ...

File: metvocab/mmdgroup.py (hunk 2 above)
Comment (johtoblan) [reply metvocab/mmdgroup.py:36]: Was originally in function called in __init__. Should we make it a method not called by __init__?

Comment (vkbo) [reply metvocab/mmdgroup.py:36]: I think all classes should behave the same way. Makes it easier to use.

Comment (johtoblan) [reply metvocab/mmdgroup.py:36]: Roger

File: tests/tools.py [hunk 17]
```diff
@@ -32,8 +34,14 @@ def writeFile(fileName, fileData):
         outFile.write(fileData)
 
 
+def readJson(fileName):
+    """Returns the content of a file as a json-object."""
```
Comment (vkbo) [inline tests/tools.py:38]: No, it returns a `dict`.

File: metvocab/mmdgroup.py [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+            if found is True:
+                Resource = self._get_resource(concept, "rdfs:seeAlso")
+                return {
+                    "Short_Name": self._get_label(concept, "prefLabel"),
+                    "Long_Name": self._get_label(concept, "altLabel"),
+                    "Resource": Resource if "wmo" in Resource else None
+                }
+        return {}
+
+    def populate(self):
```
Comment (vkbo) [inline metvocab/mmdgroup.py:55]: No, I meant do exactly the same as the other classes.

**Review by vkbo — CHANGES_REQUESTED** [review]
Two more bits

**Review by vkbo — APPROVED** [review]

**Pull request merged**